nixos/nixos-generate-config: add networkmanager to template

Artturin opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Improve the experience for users

Closes https://github.com/NixOS/nixpkgs/issues/146226
Closes https://github.com/NixOS/nixpkgs/issues/73595
###### Things done
Test with `nix build --expr 'with import ./. {}; (nixos { }).nixos-generate-config' --impure && ./result/bin/nixos-generate-config --dir config1` the files will be in ./config1 dir

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Artturin committed:
nixos/tools: add networkmanager to configuration template

Artturin commented:
```
rm -rf config1 && nix build --expr 'with import ./. {}; (nixos { }).nixos-generate-config' --impure && ./result/bin/nixos-generate-config --dir config1
writing config1/hardware-configuration.nix...
writing config1/configuration.nix...
Username without spaces. Leave empty to skip: user1
Password:
For more hardware-specific settings, see https://github.com/NixOS/nixos-hardware.
```

```
  users.users.user1 = {
    initialHashedPassword = "$6$sHK4JD8qVRLhXfW$bwymSTga2dwDYt3qF6XQ9vSz3Lq0UaX/L12Seobub3/5DW64k3sFyMFFuL37Dm9Su.EsEGwp6G25ynKDnHfHd1";
    isNormalUser = true;
    extraGroups = [ "wheel" ]; # Enable ‘sudo’ for the user.
  };
```

```
rm -rf config1 && nix build --expr 'with import ./. {}; (nixos { }).nixos-generate-config' --impure && ./result/bin/nixos-generate-config --dir config1 --uninteractive
writing config1/hardware-configuration.nix...
writing config1/configuration.nix...
For more hardware-specific settings, see https://github.com/NixOS/nixos-hardware.
```

grahamc commented:
Have you tested this when not at a TTY / non-interactive mode?

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -62,6 +65,12 @@ sub runCommand {
+    elsif ($arg eq "--dont-configure-user") {
+        $dontConfigureUser = 1;
+    }
+    elsif ($arg eq "--uninteractive") {
```
How about `--noninteractive` or `--batch`?

Artturin replied:
thanks noninteractive is better

Artturin commented:
> Have you tested this when not at a TTY / non-interactive mode?

i added logic for that and tested with
```
$ rm -rf config1 && nix build --expr 'with import ./. {}; (nixos { }).nixos-generate-config' --impure && env - HOME="$HOME" USER="$USER" PATH=$PATH /bin/sh -c './result/bin/nixos-generate-config --dir config1' </dev/null >job.log 2>&1
$ cat job.log
writing config1/hardware-configuration.nix...
writing config1/configuration.nix...
For more hardware-specific settings, see https://github.com/NixOS/nixos-hardware.
```

found in https://unix.stackexchange.com/a/23928

roberth commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -62,6 +65,12 @@ sub runCommand {
+    elsif ($arg eq "--dont-configure-user") {
+        $dontConfigureUser = 1;
+    }
+    elsif ($arg eq "--noninteractive") {
```
```suggestion
    elsif ($arg eq "--non-interactive") {
```

roberth commented:
For testing, I think you can add to the existing installer tests.
For documentation, there's nixos/doc/manual/man-nixos-generate-config.xml

Artturin commented:
https://github.com/NixOS/nixpkgs/blob/9c217dfadb74c21aa43f74c6c59881a9c173b5d4/nixos/modules/installer/tools/tools.nix#L36
this is wrong so i'll fix that too and add `mkpasswd`

SuperSamus commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -134,6 +134,7 @@ in
       $bootLoaderConfig
         # networking.hostName = "nixos"; # Define your hostname.
         # networking.wireless.enable = true;  # Enables wireless support via wpa_supplicant.
+        # networking.networkmanager.enable = true;  # Easiest to use and most distros use this by default.
```
Maybe it could be made more clear that `networking.wireless.enable` and `networking.networkmanager.enable` are mutually exclusive?

Artturin replied:
i added
```
# Pick only one of the below networking options.
```

Artturin committed:
nixos/tools: add more keymap options to configuration template

SuperSamus commented:
Speaking of user experience, could you also consider implementing #146226?
(Or should it be a separate PR?)

Artturin commented:
> Speaking of user experience, could you also consider implementing #146226? (Or should it be a separate PR?)

sure i can 

i'll do random improvements to the configs in this pr

Artturin committed:
nixos/nixos-generate-config: fix proram path replacements

btrfs-progs was made optional in 2019 c01ea27ce356e1433da01a7b46ba40c880f2f2ac
because it didn't cross-compile. it now does

Artturin committed:
nixos/nixos-generate-config: move dhcpConfig from configuration.nix to hardware-configuration.nix

Artturin commented:
> Speaking of user experience, could you also consider implementing #146226? (Or should it be a separate PR?)

done

Artturin commented:
dropped the user generation from this pr, the code for it can still be found at https://github.com/Artturin/nixpkgs/tree/nixosgenerateconfig2-orig
